### ISSUE TITLE
[only uses GPU0 while generating audio]

### ISSUE THREAD

**Issue by bilalabbasi55** [kan-bayashi/ParallelWaveGAN#259]
I tried to run the following code example on google's VM with 2GPUs.

https://colab.research.google.com/github/espnet/notebook/blob/master/espnet2_tts_realtime_demo.ipynb#scrollTo=3TTAygALqY6T

But for some reason it only uses GPU0 and GPU1 is always free. Please find the attached snapshot of the issue i am having.
Can someone please point me to right direction i am new to this world. I want to use all available GPUs to run this code.

<img width="683" alt="gpu0_only" src="https://user-images.githubusercontent.com/31213804/111613625-89e81d00-8800-11eb-8dbb-9d3334497778.png">

**Comment by kan-bayashi** [comment]
Why do you want to use multi GPU in decoding?
Do you want to decode the utterances in parallel?
If you decode only single utterance, you cannot obtain the gain.

**Comment by kan-bayashi** [comment]
The use of multiple GPU in a single process is a little bit complicated.
Multiple processes are much easier.
For example, launch API instance for each GPU, send split requests to each API process.

**Comment by bilalabbasi55** [comment]
@kan-bayashi i did same, but still i am getting out of memory issues, system had some better performance but could not do much API hits per second.
What i feel like is system garbage collection or cache clearing is not working properly (again its a guess, as i am not an expert to look and see what is missing).

I have added these two lines extra that were not present in initial code.
torch.cuda.empty_cache() //For clearing cache
and
y = wav.view(-1).detach().cpu().numpy() //for moving results to CPU.

My TTS sentence is not more than 30seconds, and i cannot make 3-4 API hits per second. Can you help me what i can do here?
Attached is the GPUs (4) we are using.
![image3](https://user-images.githubusercontent.com/31213804/112527601-0cef1180-8dc5-11eb-9166-c4694c6d2595.png)

**Comment by kan-bayashi** [comment]
Why did you launch the process on different GPU?

**Comment by bilalabbasi55** [comment]
@kan-bayashi as recommended by you
"For example, launch API instance for each GPU, send split requests to each API process."
I have lunched 4 processes on 4 different GPUs.(attached image2)
my concern is still same, its still going out of memory and raising the errors.
"Mar 25 13:20:49 tts-vm-gpu-1 ttsapi[559]: RuntimeError: CUDA out of memory. Tried to allocate 120.00 MiB (GPU 0; 14.76 GiB total capacity; 8.88 GiB already allocated; 97.75 MiB free; 9.56 GiB reserved in total by PyTorch)"

I just want to know why its getting out of memory, i am not really making more than 3-4 hits per second. 

<img width="646" alt="image2" src="https://user-images.githubusercontent.com/31213804/112594072-2eceaf80-8e2a-11eb-9830-7814f83969e9.png">

**Comment by kan-bayashi** [comment]
The required memory depends on the length of inputs.

**Comment by bilalabbasi55** [comment]
@kan-bayashi  I have 25-30 seconds average length of audio, does it really mean i have to add more GPUs?
To be honest i just want your expert opinion that its not something related to garbage collection etc.

Being naive i think 1-2 requests per second cannot make these 4GPUs(64GB) out of memory.

**Comment by kan-bayashi** [comment]
You can check the memory consumption by yourself.
You launched 4 process on a single gpu, therefore, each gpu can use around 4GB.
In that case, OOM may happen when 4 processes decodes 30 seconds audio at the same time.

**Comment by bilalabbasi55** [comment]
@kan-bayashi  I am sorry, as i could not understand that, i have launched 4 process on 4 different GPUs. i-e GPU0, GPU1, GPU2,GPU3.
It should be 16GB available for each process. Please let me know if i am doing anything wrong here.

**Comment by kan-bayashi** [comment]
From your figure, your process located on GPU, right?

**Comment by bilalabbasi55** [comment]
@kan-bayashi  I have four processes and i tired to use different GPUs for each process. PFA images for three of my process, you will see cuda:0, cuda:1, cuda:2
I know from nvidia-smi it seems like other processes are still using cuda:0 for something(sharing memory with cuda:0, it should not happen)

<img width="575" alt="Screenshot 2021-03-26 at 12 58 41 PM" src="https://user-images.githubusercontent.com/31213804/112600649-4a8a8380-8e33-11eb-8c18-d50325482d52.png">
<img width="621" alt="Screenshot 2021-03-26 at 12 58 58 PM" src="https://user-images.githubusercontent.com/31213804/112600653-4bbbb080-8e33-11eb-9b24-335733e45664.png">


<img width="687" alt="Screenshot 2021-03-26 at 1 00 13 PM" src="https://user-images.githubusercontent.com/31213804/112600632-45c5cf80-8e33-11eb-835a-40cbf35d9126.png">

**Comment by kan-bayashi** [comment]
did you move text2speech as well?

**Comment by bilalabbasi55** [comment]
@kan-bayashi I guess i have not moved that, i can attach my code here(please find the link below), can you point me how can i do that?
https://wtools.io/paste-code/b4s7

**Comment by kan-bayashi** [comment]
Maybe it is OK.
I think at first, you need to test the maximum length input with a single process.
I do not know the request will be sent in parallel or it can wait the previous task.
Anyway, your question is out of scope of this repository. Please solve by yourself.

**Comment by bilalabbasi55** [comment]
@kan-bayashi thanks a lot for getting sometime for this thread.
